### ISSUE TITLE
Persistent session across multiple connections

### ISSUE THREAD

**Issue by lgrahl** [threema-ch/threema-web#498]
We have and we always will have issues with Threema Web losing the connection to the app for several reasons:

1. The app is being suspended,
2. the network of the browser side changes,
3. the network of the app side changes, or
4. (buggy implementations).

3. and 4. could potentially be improved by supporting ICE restarts. However, making the session persistent across multiple connections would be a solution for all four items. It would also allow for dropping the permanent background connection after a period of inactivity without much of an impact which I think is a vital feature Threema Web should have. The only downside is that this will require a major protocol update along with some clever synchronisation.

**Comment by ovalseven8** [comment]
Will this also help for roaming? Currently, it takes a lot of time until a connection is re-established.

**Comment by ovalseven8** [comment]
As far as I understand, this would help a lot with roaming. Currently, it takes a lot of time until a connection is re-established.

**Comment by lgrahl** [comment]
Only Android left.

**Comment by LtPatterson** [comment]
Getting excessively high battery drain on Android 9.0 stable release. Over 20% if I leave it running for most of a workday.

**Comment by lgrahl** [comment]
@LtPatterson We are aware of general connectivity issues and the associated battery drain. If you want your individual case investigated, please contact the Threema support directly.

**Comment by ovalseven8** [comment]
Well, it's not an individual case, so there's nothing the Threema Support can do. Let's hope the persistent connection for Android and further improvements will help. :)

**Comment by lgrahl** [comment]
Note for us: We should add statistics (number of received chunks, last received chunk timestamps, number of sent chunks, last sent chunk, rx total, tx total, ...) when resolving this for Android.

**Comment by FelschR** [comment]
Any news?

**Comment by lgrahl** [comment]
Not yet but be assured this is on my top N list of things to resolve.

**Comment by lgrahl** [comment]
Another note for us: Make sure to push several times (with some backoff mechanism) in order to improve issues mentioned in #106.

**Comment by lbort** [comment]
I'm not sure this is the right issue for this comment, but it was the best I could find. 

tl;dr: Having an active and an inactive session on the phone (sometimes?) draws insane levels of power. phone gets hot and battery dies in 1h. Also: would be nice if you would consider google-free phones in your fix. 

Steps to reproduce: 
I use threema-web on an AOSP device without google play services, i.e. push notifications not available. This makes session resumption with password cumbersome, its easier to just scan a new barcode than typing the password *and* starting the session manually. This means, I don't use saved sessions but fairly often start one session, "terminate" it by just closing the browser window/PC going to sleep/whatever. Opening a new session, leads to two sessions in the list on the phone, one of them active. In this state, power consumption (sometimes) is so high that my phone gets hot. I think the abandonned session has an orange wifi-like indicator during these times. It would be awesome, if you could solve all this with something that does not rely on push notifications. Or even better: could you just use a push notification server on your own, which is used if google services are not found? That would also fix the regular app, and replace the polling option.

**Comment by ovalseven8** [comment]
When this is implemented, it would be interesting to see how much this helps with battery drain. Perhaps Threema can do some analysis and statistics on this. :)

**Comment by lgrahl** [comment]
@lbort Improving non-FCM is on our list. However, I suggest checking out [microG](https://microg.org/).

@ovalseven8 That cannot really be measured since it means moving from a persistent connection to on-demand... which depends on your demand. :slightly_smiling_face:

**Comment by Hornochs** [comment]
Hey @lgrahl,

just wanted to ask cautionly if there is any news about this issue :) 

A group of 10 switched from Whatsapp to Threema and four of us just noticed the heavy battery drain when Using Threema Web all day 😅

**Comment by lgrahl** [comment]
We're working on something new that will replace Threema Web.

In the meantime, what helps is a stable WiFi connection (less connectivity checks), using Firefox (way less aggressive keep-alive timers) and keeping the phone connected to the PC (or a slow charger) because Android and iOS are less restrictive on background activity when the phone is charging.

**Comment by Hornochs** [comment]
Thank you very much for the fast reply!

Okay, I was using it in Franz ... Then I'll switch to Firefox for that, no problem. I already had an stable WiFi Connection. Well, then I have to change my charging behavior 😃

Is there any ETA already when it will be replaced or where I'm able to subscribe for updates?

**Comment by lgrahl** [comment]
I can't provide an ETA but you can subscribe to the [Threema Channel](https://threema.id/%2ATHREEMA?text=info) where any news would certainly be announced.

**Comment by Hornochs** [comment]
Hi @lgrahl
I noticed what causes the high battery drain, that was not Threema Web alone. As soon as I'm at work and using it the service mdnsd is using quadruble the battery then Threema Web. At Home I don't have this. I guess it's because at work I'm on a different network (Guest WiFi) then my PC using it. 🤔

**Comment by lgrahl** [comment]
What are you referring to here? Battery drain on the smartphone or the PC running Threema Web? What kind of system does it run?

**Comment by Hornochs** [comment]
On the Smartphone and the battery drain:

![image](https://user-images.githubusercontent.com/63955136/105027550-32237480-5a50-11eb-9b39-6f831c96c519.png)

This high drain from mdnsd is only in my office. At home it's not nearly as 5% drain.

**Comment by lgrahl** [comment]
Very interesting. Can you reproduce this behaviour when WiFi is off?

**Comment by Hornochs** [comment]
Very interesting, I just switched to used mAh and see whats happening with WiFi Disabled:

![image](https://user-images.githubusercontent.com/63955136/105056474-86d8e680-5a74-11eb-8a56-06ac7a480c77.png)

While beeing WiFi enabled this happens:

![image](https://user-images.githubusercontent.com/63955136/105056520-95bf9900-5a74-11eb-9da3-0ed491b602eb.png)

Super weird 😅

**Comment by lgrahl** [comment]
I have a hunch what this might be. mDNS is used by WebRTC as a privacy measure. I'll see if I can investigate this at some point.

**Comment by lgrahl** [comment]
@Hornochs The question now would be: Does the mdns daemon also go berserk when you're not using Threema Web in that network?

**Comment by Hornochs** [comment]
> @Hornochs The question now would be: Does the mdns daemon also go berserk when you're not using Threema Web in that network?

Thats true, While not using Threema Web mdnsd stays silent.

**Comment by HWiese1980** [comment]
How's progress going here? There's the new Desktop App which appears to be causing the same battery drain on my Android phone.

**Comment by lgrahl** [comment]
Threema Web is in maintenance mode while we are working on a new solution that should resolve some of the long-standing issues we were having with Threema Web, including this one. Threema Desktop 1.x is based on Threema Web, so the same applies to Threema Desktop 1.x.

**Comment by johseg** [comment]
so what is the state here? It's been two years and with my new Pixel 7a I can't get through the day on my battery charge when using Threema web vs. having 60% left at the end of the day when I don't use it. 

I like Threema and pushed my whole family to use it, but this has been a problem for many, many years now and seems to only get worse. If there's no solution for this we'll likely move to Signal

**Comment by threema-lenny** [comment]
The solution is Multi-Device and [Threema Desktop 2.0](https://threema.ch/en/download-md). Development for Android is currently ongoing.
